### PR TITLE
CTE deprecation

### LINKED ISSUE #166: Automatically inferring dependencies in CTE feature
Note: this is blocked by https://github.com/ploomber/jupysql/issues/158 - once that's merged, we can use sqlglot to parse SQL and find which tables a given SQL query is using.

**Important:** our guest blog posts might use the `--with` feature, so if they use it, we'll need to update them once we remove `--with`

## improving the CTE feature ##

we should deprecate `--with` and automatically infer what are the dependencies for a given query.

```python
%load_ext sql
%sql duckdb://
```

for example, if a user runs this:

```python
%%sql
select *
from some_table
```

**Console output (1/1):**

```txt
*  duckdb://
(duckdb.CatalogException) Catalog Error: Table with name some_table does not exist!
Did you mean "pg_tables"?
LINE 2: from some_table
             ^
[SQL: select *
from some_table]
(Background on this error at: https://sqlalche.me/e/14/f405)
```

we should parse the sql query, detect that this is querying a `some_table` entity. Then check which queries the user has stored and if we find a match, we should automatically add the definition to form the CTE (currently, the user needs to pass `--with some_table` manually)

this introduces a side-effect problem where users might run`--save {something}` where `{something}` is also table in the database; hence masking the table and leading to all `select * from something` querying the snippet instead of the actual table. To alleviate this, we could print something like:

> Forming CTE from saved snippets: A, B, C, and D

So it's clear that these are coming from the locally defined snippets, not from tables. Then, if a user wants to delete a definition (so it no longer masks the table), we could add a command like this:

```
%sqlcmd snippets something --delete
```

Keep in mind that there might be snippets that use something as definition, in such case, we should display a warning:

> "A", "B", "C" depend on "something". Pass --delete-force to only delete "something", pass --delete-force-all to delete "A", "B", "C", and "something"

Finally, we can offer typo resolution. For example, If I try to run this:

```python
%%sql
select * from somethingo
```

and something throws an error. we could check the stored snippets and see if there are any [close matches](https://docs.python.org/3/library/difflib.html#difflib.get_close_matches), and if so, append something like this to the error raised by the DB driver:

> There is a stored snippet called "something". Did you mean to use that one?

### PR DESCRIPTION
**Pull request by neelasha23**
## Describe your changes
Deprecate --with when forming CTEs

## Issue number

Closes #166 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--486.org.readthedocs.build/en/486/

### CODE REVIEW

**Comment by neelasha23** [comment]
Covered most of the test cases I could think of. Still testing to check for bugs. Meanwhile, one round of review would be helpful.

@edublancas @idomic 

Also, there is a circular dependency issue whenever `util.py` is imported in some of the modules. So had to create separate module, e.g., `query_util.py` for now.

```
ImportError while loading conftest '/Users/neelashasen/Dev/jupysql/src/tests/conftest.py'.
src/tests/conftest.py:8: in <module>
    from sql.magic import SqlMagic, RenderMagic
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/__init__.py:1: in <module>
    from .magic import RenderMagic, SqlMagic, load_ipython_extension
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/magic.py:22: in <module>
    import sql.connection
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/connection.py:11: in <module>
    from sql.store import store
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/store.py:12: in <module>
    from sql import util
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/util.py:1: in <module>
    from sql import inspect
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/inspect.py:4: in <module>
    from sql.connection import Connection
E   ImportError: cannot import name 'Connection' from partially initialized module 'sql.connection' (most likely due to a circular import) (/Users/neelashasen/miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/connection.py)
```

File: doc/api/magic-snippets.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+:tags: [hide-output]
+
+%%sql --save chinstrap_sub
+SELECT * FROM chinstrap where island == 'Dream'
+```
++++
+
+Trying to delete `chinstrap` snippet using `%sqlcmd snippets -d chinstrap` will display an error message:
+
+![error](../static/snippet-delete-error.png)
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply doc/api/magic-snippets.md:102]: It was causing the docs build to fail so I added the image instead.

File: src/tests/test_magic_plot.py [hunk 2]
```diff
@@ -63,7 +64,6 @@ def test_validate_arguments(tmp_empty, ip, cell, error_type, error_message):
         "%sqlplot boxplot --table data.csv --column x",
         "%sqlplot box --table data.csv --column x",
         "%sqlplot boxplot --table data.csv --column x --orient h",
-        "%sqlplot boxplot --table subset --column x --with subset",
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/tests/test_magic_plot.py:66]: No we wouldnt support `--with`. There's another testcase for warning

File: doc/api/magic-snippets.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (yafimvo) [reply doc/api/magic-snippets.md:102]: We can use `:tags: [raises-exception]`

[reference](https://github.com/ploomber/jupysql/blob/master/doc/community/developer-guide.md?plain=1#L177)

File: src/sql/magic.py [hunk 3]
```diff
@@ -436,6 +453,18 @@ def interactive_execute_wrapper(**kwargs):
                 if detailed_msg is not None:
                     err = exceptions.UsageError(detailed_msg)
                     raise err
+                    # TODO: move to error_messages.py
```
Comment (edublancas) [inline src/sql/magic.py:456]: looks like there is some pending work here

File: src/sql/magic.py [hunk 4]
```diff
@@ -436,6 +453,18 @@ def interactive_execute_wrapper(**kwargs):
...
+                    tables = query_util.extract_tables_from_query(command.sql)
+                    for table in tables:
+                        suggestions = get_close_matches(table, list(self._store))
+                        if len(suggestions) > 0:
+                            err_message = f"There is no table with name {table!r}."
+                            suggestions_message = get_suggestions_message(suggestions)
+                            raise exceptions.TableNotFoundError(
+                                f"{err_message}{suggestions_message}"
+                            )
+                    print(e)
```
Comment (edublancas) [inline src/sql/magic.py:467]: I was unable to get this working (ran the compose.md file):

image


changed from genres_fav -> genres_fav2

File: src/sql/magic_cmd.py [hunk 5]
```diff
@@ -210,6 +211,81 @@ def execute(self, cmd_name="", others="", cell="", local_ns=None):
 
             return report
 
+        elif cmd_name == "snippets":
+            parser = CmdParser()
+            parser.add_argument(
+                "-d", "--delete", type=str, help="Delete stored snippet", required=False
+            )
+            parser.add_argument(
+                "-fd",
```
[the remark this answers is not in the document]
Comment (edublancas) [reply src/sql/magic_cmd.py:220]: we should be consistent with arguments naming. we're doing capital arguments instead of two lowercase ones. so this should be `-D`

File: src/sql/magic_cmd.py [hunk 6]
```diff
@@ -210,6 +211,81 @@ def execute(self, cmd_name="", others="", cell="", local_ns=None):
...
+                "--delete-force",
+                type=str,
+                help="Force Delete stored snippet",
+                required=False,
+            )
+            parser.add_argument(
+                "-da",
```
Comment (edublancas) [inline src/sql/magic_cmd.py:227]: `-da` -> `-A`

File: src/sql/magic_cmd.py [hunk 7]
```diff
@@ -210,6 +211,81 @@ def execute(self, cmd_name="", others="", cell="", local_ns=None):
...
+            SNIPPET_ARGS = [args.delete, args.delete_force, args.delete_force_all]
+            if SNIPPET_ARGS.count(None) == len(SNIPPET_ARGS):
+                return ", ".join(util.get_all_keys())
+            if args.delete:
+                deps = util.get_key_dependents(args.delete)
+                if deps:
+                    deps = ", ".join(deps)
+                    raise exceptions.UsageError(
+                        f"The following tables are dependent on {args.delete}: {deps}."
+                        f"Pass --delete-force to only delete {args.delete}."
```
[the remark this answers is not in the document]
Comment (edublancas) [reply src/sql/magic_cmd.py:243]: agree. I had a hard time understanding the error message because of this

File: src/sql/magic_cmd.py [hunk 8]
```diff
@@ -210,6 +211,81 @@ def execute(self, cmd_name="", others="", cell="", local_ns=None):
 
             return report
 
+        elif cmd_name == "snippets":
```
Comment (edublancas) [inline src/sql/magic_cmd.py:214]: this if statement is becoming longer and longer. please move the logic into a different file: sql/sqlcmd.py so the if statement becomes:

```python
elif cmd_name == "snippets":
    sqlcmd_snippets()
```

File: src/sql/magic_cmd.py [hunk 9]
```diff
@@ -210,6 +211,81 @@ def execute(self, cmd_name="", others="", cell="", local_ns=None):
...
+                        )
+                    else:
+                        display_msg = (
+                            f"{display_msg}. There are no more saved snippets."
+                        )
+                    return display_msg
+            elif args.delete_force:
+                key = args.delete_force
+                deps = util.get_key_dependents(key)
+                warnings.warn(
```
Comment (edublancas) [inline src/sql/magic_cmd.py:263]: this should not be a warning, but a print statement (it's a confirmation message)

there are two cases:

- if the deleted snippet didn't have any dependencies just say "{snippet} has been deleted." (past tense)
- if there are dependencies say: "{snippet} has been deleted. {a}, {b} depend on it."

File: src/sql/magic_cmd.py [hunk 10]
```diff
@@ -210,6 +211,81 @@ def execute(self, cmd_name="", others="", cell="", local_ns=None):
...
+                return display_msg
+
+
+def _modify_display_msg(msg, keys):
+    if keys:
+        display_msg = f"{msg}. Current saved snippets : {', '.join(keys)}"
+    else:
+        display_msg = f"{msg}. There are no more saved snippets."
+    return display_msg
+
```
Comment (edublancas) [inline src/sql/magic_cmd.py:288]:
```suggestion
def _modify_display_msg(msg, keys):
    if keys:
        display_msg = f"{msg}. Stored snippets : {', '.join(keys)}"
    else:
        display_msg = f"{msg}. There are no stored snippets"
    return display_msg

```

File: src/sql/magic_plot.py [hunk 11]
```diff
@@ -73,23 +74,40 @@ def execute(self, line="", cell="", local_ns=None):
         column = util.sanitize_identifier(column)
         table = util.sanitize_identifier(cmd.args.table)
 
+        if cmd.args.with_:
+            warnings.warn(
+                "--with will be deprecated soon, and dependencies "
+                "will be inferred from saved queries instead",
+                FutureWarning,
+            )
```
Comment (edublancas) [inline src/sql/magic_plot.py:82]: I suggested changing this warning message so let's move it to a single place and call it from there:

e.g., 
```python
        if cmd.args.with_:
            show_with_deprecation_warning()
```

File: src/sql/store.py [hunk 12]
```diff
@@ -132,6 +147,15 @@ def _get_dependencies(store, keys):
...
+def _get_dependents_for_key(store, key):
+    key_dependents = []
+    for k in list(store):
+        deps = _get_dependencies_for_key(store, k)
+        if key in deps:
+            key_dependents.append(k)
+    return key_dependents
+
+
 def _get_dependencies_for_key(store, key):
```
[the remark this answers is not in the document]
Comment (edublancas) [reply src/sql/store.py:159]: this is old code, hasn't been modified by this PR

File: src/tests/conftest.py [hunk 13]
```diff
@@ -94,10 +94,10 @@ def ip(ip_empty):
         ],
     )
     yield ip_empty
-    runsql(ip_empty, "DROP TABLE test")
-    runsql(ip_empty, "DROP TABLE author")
-    runsql(ip_empty, "DROP TABLE website")
-    runsql(ip_empty, "DROP TABLE number_table")
+    runsql(ip_empty, "DROP TABLE IF EXISTS test")
```
Comment (edublancas) [inline src/tests/conftest.py:97]: why was this changed? if we're always creating a new DB, there's a guarantee that the tables exist there (maybe it's ok to add IF EXISTS but I'm wondering why)

File: src/tests/test_compose.py [hunk 14]
```diff
@@ -7,12 +11,129 @@ def test_compose(ip):
...
+        ip.run_cell_magic(
+            "sql",
+            "--with author_sub --save final",
+            "SELECT last_name FROM author_sub;",
+        )
+        assert len(record) == 1
+        assert (
+            "--with will be deprecated soon, and dependencies "
+            "will be inferred from saved queries instead" in record[0].message.args[0]
+        )
```
Comment (edublancas) [inline src/tests/test_compose.py:45]: the assert statements should outside of the `with` context manager

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/sql/magic_cmd.py [hunk 15]
```diff
@@ -210,6 +211,81 @@ def execute(self, cmd_name="", others="", cell="", local_ns=None):
...
+                remaining_keys = util.del_saved_key(key)
+                display_msg = f"Deleted snippet : {key}"
+                display_msg = _modify_display_msg(display_msg, remaining_keys)
+                return display_msg
+
+            elif args.delete_force_all:
+                deps = util.get_key_dependents(args.delete_force_all)
+                deps.append(args.delete_force_all)
+                display_msg = f"Deleted snippets : {', '.join(deps)}"
+                for key in deps:
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/sql/magic_cmd.py:276]: in `delete-force-all` there won't be any more dependent snippets left

File: src/sql/query_util.py [hunk 16]
```diff
@@ -0,0 +1,10 @@
+from sqlglot import parse_one, exp
+from sqlglot.errors import ParseError
+
+
+def extract_tables_from_query(query):
+    try:
+        tables = [table.name for table in parse_one(query).find_all(exp.Table)]
+        return tables
+    except ParseError:
+        return []
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/sql/query_util.py:10]: It means sqlglot was not able to detect any tables in the query, can happen if there are syntax issues

[the remark this answers is not in the document]
Comment (yafimvo) [reply src/sql/query_util.py:10]: So if the table exists and there is a syntax error we return an empty array? 
I would expect `extract_tables_from_query` to throw an error if there is a syntax issue but if that's not possible I think we should document it.

File: src/sql/__init__.py [hunk 17]
```diff
@@ -11,4 +12,5 @@
     "load_ipython_extension",
     "SYNTAX_ERROR",
     "PLOOMBER_DOCS_LINK_STR",
+    "del_all_saved_keys",
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/sql/__init__.py]: added this as a fixture in test file

File: src/tests/test_compose.py (hunk 14 above)
Comment (neelasha23) [reply src/tests/test_compose.py]: removed since it is print statements now instead of warning

File: src/sql/magic_cmd.py (hunk 8 above)
Comment (neelasha23) [reply src/sql/magic_cmd.py:219]: Should I move the logic for the other command as well? e.g., tables/columns

File: src/sql/magic.py (hunk 4 above)
Comment (neelasha23) [reply src/sql/magic.py:463]: It would be `ON t.GenreId = genres_fav2.GenreId` (It's `genres_fav.GenreId` in the above query)

Screenshot 2023-05-26 at 1 20 53 PM

@edublancas

File: src/sql/magic.py (hunk 3 above)
Comment (neelasha23) [reply src/sql/magic.py:452]: I'm not able to move this logic to `error_messages.py` because of the circular dependency mentioned [here](https://github.com/ploomber/jupysql/pull/486#issuecomment-1559645056)

File: src/tests/conftest.py (hunk 13 above)
Comment (neelasha23) [reply src/tests/conftest.py]: Without the `IF EXISTS` clause I see this error. This is coming up because now there's a logic to check if there are stored snippets similar to the one in query (typo case). This might be happening because in some testcase the  test tables are not getting created properly but then it's going back to delete it. I'm not sure exactly which testcase this happens in , also it doesn't fail on my local.

Screenshot 2023-05-26 at 5 43 57 PM

File: src/sql/magic_cmd.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/sql/magic_cmd.py]: Warning has been replaced with print statements now.

**Comment by tonykploomber** [comment]
By the way, do we need to [take telemetry into this](https://github.com/ploomber/jupysql/issues/166#issuecomment-1514969763)?

File: src/sql/query_util.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/sql/query_util.py:10]: This is an example of query where table exists (as stored snippet) but there's a syntax error in the query. It'll throw error:

Screenshot 2023-05-29 at 12 56 54 PM

@yafimvo

[the remark this answers is not in the document]
Comment (yafimvo) [reply src/sql/query_util.py:10]: Great. 

Since we assume the query is valid, I think we should add a note that this function (`extract_tables_from_query`) doesn't throw errors if the query is invalid.

If someone will want to use this function in the future they should know that if it returns `[]` - it doesn't mean there is no such table.

example:

```
query_util.extract_tables_from_query("select * from table_name limi 2")
```

The table exists but it will return `[]`

P.S
Another option is to check if the query is valid inside `extract_tables_from_query` and throw an error.

File: src/tests/test_magic_cmd.py [hunk 18]
```diff
@@ -263,3 +264,127 @@ def test_test_error(ip, cell, error_type, error_message):
...
+    assert isinstance(out.error_in_exec, UsageError)
+    assert (
+        str(out.error_in_exec) == "The following tables are dependent on high_price: "
+        "high_price_a, high_price_b.Pass --delete-force to only "
+        "delete high_price.Pass --delete-force-all to delete "
+        "high_price_a, high_price_b and high_price"
+    )
+    # Clean saved keys
+    del_all_saved_keys()
+
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/tests/test_magic_cmd.py:384]: There are some testcases for alldelete, force-delete, force-delete-all. These are applicable for snippets only not tables.
Can u give an example of snippet/table combination.
@yafimvo

**Comment by neelasha23** [comment]
> Added some comments
> 
> General notes
> 
> 1. I think we can test the util functions as well (e.g. `get_key_dependents`, `del_saved_key`, `del_all_saved_keys`)
> 2. There is missing documentation for some of the functions

Added unit tests and docstrings

File: src/sql/magic.py (hunk 3 above)
Comment (edublancas) [reply src/sql/magic.py:483]: ok, can you open an issue? we need to do some refactoring here.

I was unable to find a fix because the traceback in the link you shared is outdated, seems like some imports already changed, so let's create an issue and I can review this later.

File: src/sql/magic_cmd.py (hunk 8 above)
Comment (edublancas) [reply src/sql/magic_cmd.py:219]: just move the logic from this query, I opened an issue to move the other ones

**Review by edublancas — CHANGES_REQUESTED** [review]
minor comments

**Comment by edublancas** [comment]
there's no need to add telemetry because this isn't adding any new options (it's automatically executed whenever a SQL snippet runs)

File: src/sql/magic.py (hunk 3 above)
Comment (neelasha23) [reply src/sql/magic.py:484]: Opened [issue](https://github.com/ploomber/jupysql/issues/545)

File: src/sql/query_util.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/sql/query_util.py:10]: Currently if I try to raise an exception some fixtures are failing. Added some context in [this issue](https://github.com/ploomber/jupysql/issues/546).
For now, I'll add some comments

@yafimvo

File: src/tests/test_magic_cmd.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/tests/test_magic_cmd.py:367]: Adding discussion notes here:
```
%%sql --save gentoo
SELECT * FROM penguins.csv where species == 'Gentoo'
```

```
%%sql --save gentoo_male
SELECT * FROM gentoo where sex == 'MALE'
```

```
%%sql --save gentoo_male_body_mass
SELECT * FROM gentoo_male where body_mass_g > 5200
```

we now can try :

1. %sqlcmd snippets --delete-force gentoo_male -> make sure gentoo and gentoo_male_body_mass exist, and gentoo_male deleted. (In this case, if we run `%sql select * from gentoo_male_body_mass` we should get `UsageError: "gentoo_male" is not a valid snippet identifier` )
2. %sqlcmd snippets --delete-force-all gentoo_male -> make sure gentoo exists; and gentoo_male and gentoo_male_body_mass  deleted.
3. %sqlcmd snippets -d gentoo /gentoo_male/gentoo_male_body_mass 
4. %sqlcmd snippets --delete-force no_such_table
5. %sqlcmd snippets --delete-force-all no_such_table



Added the missing tests (Level 3 dependent table, and force-delete, force-delete-all on non-existent table)

@yafimvo

**Review by tonykploomber — APPROVED** [review]
LG

**Review by yafimvo — CHANGES_REQUESTED** [review]
added minor notes. besides that lgtm.

File: src/tests/test_extract_tables.py [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+)
+def test_extract(query, expected):
+    tables = extract_tables_from_query(query)
+    assert expected == tables
+
+
+def test_invalid_query():
+    query = "SELECT city frm Customers"
+    tables = extract_tables_from_query(query)
+    assert [] == tables
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/tests/test_extract_tables.py:85]: No we are testing the `extract_tables_from_query` function directly here. It returns `[]` if no tables extracted. As mentioned on the discussion raising exception here causes [this issue](https://github.com/ploomber/jupysql/issues/546).
UsageError will get raised when the query runs on the database

File: doc/compose.md [hunk 20]
```diff
@@ -27,7 +27,8 @@ pip install jupysql matplotlib
 ```
 
 
-*New in version 0.4.3*
+```{versionchanged} 0.8.6
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply doc/compose.md]: Changed to 0.7.7

File: src/sql/magic.py (hunk 3 above)
Comment (neelasha23) [reply src/sql/magic.py:485]: Added issue numbers

File: src/tests/test_extract_tables.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (yafimvo) [reply src/tests/test_extract_tables.py:85]: are we testing `UsageError("looks like there is a syntax error...")`? If no, I think we can add it.

File: CHANGELOG.md [hunk 21]
```diff
@@ -8,6 +8,7 @@
 * [Feature] `%sql --connections` now displays an HTML table in Jupyter and a text-based table in the terminal
 * [Fix] Fix CTE generation when the snippets have trailing semicolons
 * [Doc] Added Howto documentation for enabling JupyterLab cell runtime display ([#448](https://github.com/ploomber/jupysql/issues/448))
+* [Feature] Automated dependency inference when creating CTEs. `--with` is now deprecated and will display a warning. (#166)
```
Comment (edublancas) [inline CHANGELOG.md:11]: we released 0.7.7 today, please rebase and move to 0.7.8dev

**Review by edublancas — DISMISSED** [review]
just rebase, update the changelog entry and fix merge conflicts
